### PR TITLE
Rails associations in a working MVC application

### PR DESCRIPTION
**Pull request by gabrie-lhilarion**
# Implemented a working MVC application
- Created model, views, and controllers
- Developed a custom and yet very functional authentication (login and account creation system)
- Added some nice user interface using bootstrap

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -1,3 +1,11 @@
 Rails.application.routes.draw do
   # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
+root "events#index" 
+resources :users
+resources :sessions
+resources :events
```
Comment (lekegitrepo) [inline config/routes.rb:6]: - **_routes only for used actions, for example: `resources :events, only: [:new, :create, :show, :index]`_**

**Review by lekegitrepo — CHANGES_REQUESTED** [review]
## Status: Request changes :recycle:
#### Hi @gabrie-lhilarion :clap: :clap:
#### You have done a good :+1: However, some requirements are missing/not properly implemented. Check my comments below.

#### Tests

* [ ] **_Models should have unit tests for validations and associations (Rspec) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/unit_tests_requirement.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49723079/))_**

* [ ] **_authentication should have integration tests (Rspec + Capybara) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/integration_tests_requirement.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49723079/))_**

* [ ] **_Event managment features should have integration tests (Rspec + Capybara) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/integration_tests_requirement.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49723079/))_**

#### Error!
- **_When I tried to create an account this error was displayed after clicking the "create account" button_**
Screenshot from 2020-11-05 10-13-09

#### App Functionality
- **_The user's show page should contain events that the user is attending separated in past and upcoming **(covered by automated tests)** ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))_**

#### _Reviewed by Temitayo Adeleke_

**Comment by gabrie-lhilarion** [comment]
## WORKED ON REQUIRED CHANGES
- Created test for  model relationships
- Created integration test, for user interactions using Rspec and Capybara

File: app/models/invitation.rb [hunk 2]
```diff
@@ -0,0 +1,4 @@
+class Invitation < ApplicationRecord
+  belongs_to :user
+  belongs_to :event
+end
```
Comment (blackpintz) [inline app/models/invitation.rb:4]: Write unit tests for the `model/invitation.rb`.` The invitation_spec.rb` is empty.

File: app/models/user.rb [hunk 3]
```diff
@@ -0,0 +1,7 @@
+class User < ApplicationRecord
+  validates :first_name, :last_name, presence: true
+  validates :email, :username, presence: true, uniqueness: true
+  has_many :events, foreign_key: 'creator_id'
+  has_many :invitations
+  has_many :attended_events, through: :invitations, source: :event
+end
```
Comment (blackpintz) [inline app/models/user.rb:7]: Write unit tests for the `model/user.rb`. The ` user_spec.rb` is empty.

File: spec/features/login_user_spec.rb [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    expect(page).to have_content 'You have been logged out'
+
+    visit login_path
+    fill_in 'username', with: 'hero'
+    fill_in 'email', with: 'hero@gmail.com'
+    click_button 'login'
+
+    expect(page).to have_content 'Logged in'
+  end
+end
```
Comment (blackpintz) [inline spec/features/login_user_spec.rb:25]: This test is failing.

File: spec/features/create_user_spec.rb [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+    visit new_user_path
+
+    fill_in 'user_username', with: 'hero1one'
+    fill_in 'user_first_name', with: 'fristhero1'
+    fill_in 'user_last_name', with: 'lasthero'
+    fill_in 'user_email', with: 'user@gmail.com'
+    click_button 'create_user'
+    expect(page).to have_content 'successfull'
+  end
+end
```
Comment (blackpintz) [inline spec/features/create_user_spec.rb:14]: This test is failing.

File: spec/requests/events_request_spec.rb [hunk 6]
```diff
@@ -0,0 +1,4 @@
+require 'rails_helper'
+
+RSpec.describe 'Events', type: :request do
+end
```
Comment (blackpintz) [inline spec/requests/events_request_spec.rb:4]: Add tests here.

File: spec/requests/sessions_request_spec.rb [hunk 7]
```diff
@@ -0,0 +1,4 @@
+require 'rails_helper'
+
+RSpec.describe 'Sessions', type: :request do
+end
```
Comment (blackpintz) [inline spec/requests/sessions_request_spec.rb:4]: Add tests here.

File: spec/requests/users_request_spec.rb [hunk 8]
```diff
@@ -0,0 +1,4 @@
+require 'rails_helper'
+
+RSpec.describe 'Users', type: :request do
+end
```
Comment (blackpintz) [inline spec/requests/users_request_spec.rb:4]: Add tests here.

**Review by blackpintz — CHANGES_REQUESTED** [review]
# Status: Changes Required

## Hello team,

**Great job implementing the changes. However, there are a few fixes needed.**

1). **The application shows the error below when I try to view any of the routes.**

route-err.

2). **Please check more in the comments below.**

### Happy coding 😃.

**PS: You can reach me on slack (Rose Wanjohi).**

**Review by acushlakoncept — CHANGES_REQUESTED** [review]
# Status: Changes Required ♻️

### Excellent Job Team 👏  

You've done an awesome job so far 👌. But there are still some changes that are necessary for this project to be approved.

It looks like you broke some things while fixing previous errors.

- [ ] I get an error when I click **See all Events** 
Screen Shot 2020-11-14 at 1 38 00 PM

- [ ] When trying to create an event, it throws an error: 
Screen Shot 2020-11-14 at 1 52 15 PM

### use `rails routes` to be certain of the routes you are using in your view.


- [ ] Clicking on Edit button for users throws an error: _The action 'edit' could not be found for UsersController_

- [ ] There should be absolutely NO logic in views. >Views are only for displaying information that becomes available to it from its corresponding controller not generating it. **Resources**
* [Where Do I Put My Code?](http://codefol.io/posts/Where-Do-I-Put-My-Code)
* [The Beginner's Guide to Rails Helpers](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers)
*[Rails view helpers best practices guide](https://medium.com/@acushla4real/rails-view-helpers-best-practices-guide-d7d7b6db3db4)
* [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from Pathwright View-related logic can be encapsulated in View helpers (app/helpers/*_helper.rb)

- [ ] [OPTIONAL] When a user clicks on `Attend this event`, it should change the text to something like `Attending`. And if the user clicks a second time, update the text to suit the case. 'You're already an attendee' or something instead of 'You are already a member. You can only register once!' This is purely optional.

- [ ] Since Testing has been included, update your readme file to show how to run tests on your application.

- [ ] Your tests are failing, kindly fix it.

### Cheers and Happy coding! 💻 👏

(Feel free to contact me on slack(@acushla), if you have any additional question)
Uduak Essien
[Twitter](https://twitter.com/acushlakoncept) | [LinkedIn](https://www.linkedin.com/in/acushlakoncept/)

**Review by mcrd25 — APPROVED** [review]
# Changes Approved :heavy_check_mark: 

Hi @gabrie-lhilarion :wave: 

Good job, 

You addressed all points of lass review, so there is no need for further review.
However, please consider my optional comments:

### Recommendations
- Maybe the Welcome label on the navbar should not be a button, a button represents completing an action which that does not do. Screenshot from 2020-11-18 10-29-06

- As was mentioned before, it is not user-friendly to have the text say attend this event after the user has already attended the event. Consider removing it or replacing it with some kind of unattended functionality when the user has already clicked attended an event. 
Screenshot from 2020-11-18 10-29-12


Hope this review was helpful, and congratulations! :partying_face: 

***If you have any questions, now or in the future, do not hesitate to contact me on social media, email or on slack and I will be glad to chat with you.***

*Happy coding 😄,*
**[--Maya Douglas](mayacdouglas.com)**
[GitHub](https://github.com/mcrd25) [Twitter](https://twitter.com/mcrd25) [linkedin](https://www.linkedin.com/in/mayadouglas/) [GitLab](https://gitlab.com/mcrd25)

**Pull request merged**